### PR TITLE
Save in EndTraining only if in last rank 

### PR DESCRIPTION
**Pull request by jupvfranco**
In EndTraining, the graph is saved with WITH_UPDATED_WEIGHTS and WITH_UPDATED_WEIGHTS_AND_LOSS_FUNC options, which will try to ConfigureLossFunction. However, if the graph is partitioned, only the subgraph running in the last device should have the loss function (training_runner.cc:125).

If we use pipeline parallelism to train mnist, for example, we get the error "ConfigureLossFunctionInternalEither loss function information should be provided or an external loss name should be given." in line training_session.cc:439. If we use the code in branch xuzhu/dist_cpu, we can reproduce this error.

I've also initialized two bool variables in mnist/main.cc just to prevent future odd behaviours. Happy to move this initialization to another PR though.

### CODE REVIEW

**Review by xzhu1900 — APPROVED** [review]
:shipit:
